### PR TITLE
[corlib] Implements CryptoConfig::AddAlgorithm on all profiles

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

**Comment by marek-safar** [comment]
@monojenkins backport 2018-04

**Comment by marek-safar** [comment]
@monojenkins backport 2018-02

File: mcs/class/corlib/System.Security.Cryptography/CryptoConfig.common.cs [hunk 1]
```diff
@@ -29,9 +29,43 @@
...
+		{
+			if (algorithm == null)
+				throw new ArgumentNullException (nameof (algorithm));
+			if (!algorithm.IsVisible)
+				throw new ArgumentException ("Algorithms added to CryptoConfig must be accessable from outside their assembly.", nameof (algorithm));
+			if (names == null)
+				throw new ArgumentNullException (nameof (names));
+
+			var algorithmNames = new string [names.Length];
+			Array.Copy (names, algorithmNames, algorithmNames.Length);
```
Comment (akoeplinger) [inline mcs/class/corlib/System.Security.Cryptography/CryptoConfig.common.cs]: I don't understand why we're copying to a new array here?

Comment (marek-safar) [reply mcs/class/corlib/System.Security.Cryptography/CryptoConfig.common.cs]: This is to match .net. I guess they do it in case the input array changes

Comment (akoeplinger) [reply mcs/class/corlib/System.Security.Cryptography/CryptoConfig.common.cs]: I don't really see how the input array can change here? But if .net does it then ¯\\_(ツ)_/¯

Comment (marek-safar) [reply mcs/class/corlib/System.Security.Cryptography/CryptoConfig.common.cs]: https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/CryptoConfig.cs#L305-L306

Hopefully, we get better implementation once we integrate this fully

**Pull request merged**